lib/telemetry_metrics.ex

```diff
@@ -454,4 +494,9 @@ defmodule Telemetry.Metrics do
   defp metadata_spec_to_function([]), do: fn _ -> %{} end
   defp metadata_spec_to_function(keys) when is_list(keys), do: &Map.take(&1, keys)
   defp metadata_spec_to_function(fun), do: fun
+
+  @spec measurement_spec_to_function(measurement()) ::
+          (:telemetry.event_measurement() -> number())
+  defp measurement_spec_to_function(fun) when is_function(fun, 1), do: fun
+  defp measurement_spec_to_function(term), do: &Map.get(&1, term, 0)
```

Comment:
So Ecto does not always include some times (such as the queue_time if the connection is already check out) and having a default of 0 is going to skew averages.

Maybe we should allow it to return nil, which means the measurement should be discarded, but I am also concerned that we are adding logic to all metrics and pushing complexity to all handlers.

For example, now that we have multiple measurements per event, a naive implementation can subscribe to event per measurements, while the best approach would be to unify that.

One option is to introduce something like Telemetry.Metrics.subscribe_all that will do the subscription for reporters and call the reporter with measurements, metadata and tags already extracted out but I am not sure that's the best way to go.

For now, I think we can do two things:

  1. Keep the measurement as a function or as an atom
  2. Write some explicit documentation about how to implement reporters, spitting the docs in two sections: one geared towards users of telemetry metrics and another about reporter authors.

We can do 2. in a separate commit. Thoughts?

Reply:
> * Keep the measurement as a function or as an atom

But do we keep the default as 0 in this case? 

> 2\. Write some explicit documentation about how to implement reporters, spitting the docs in two sections: one geared towards users of telemetry metrics and another about reporter authors.

I agree, we definitely need this.

Reply:
> But do we keep the default as 0 in this case?

In this case, I would not "publish anything".

Reply:
But we need to return something when measurements function is called, so probably `nil`.

Reply:
@arkgil right, that's why I am proposing for now to leave it as an atom or a function and the reporter has to handle both (for now).

Reply:
Ah, okay, I didn't get this part. Makes sense.

Reply:
Although it needs to be term, since we don't enforce atoms in measurements map.

Reply:
Ok. :) I think we can perfectly assume people are not using functions as measurements keys, so we should be fine!

Reply:
Sounds like you went with a default of `nil`, right?

That's ideal, not only does `0` skew averages, but it makes it difficult to differentiate between an actual `0` measurement and "not-reporting", which have very different meanings.

Reply:
Actually we went with "whatever the user provides". If the `:measurement` is a 1-arity function, the reporter is supposed to call it on event measurements and used a return value as the value for metric.

Take a look at StatsD reporter here for example: https://github.com/arkgil/telemetry_metrics_statsd/pull/1/files#diff-39779d31de6a0d7b26a8ddaa2b8e8f47R112.

This is not ideal, since it puts some more work on the reporter, but on the other hand, it's easier to differentiate between no-value and 0.

Reply:
One of the things we will eventually have to discuss is what happens when the event has bad measurements. For example, the measurement is meant to be a number but is nil or a struct or similar. It is probably yet another responsibility of the reporter.

Reply:
Yes, only the reporter can decide what value is valid for the system it interfaces with.